text/0000-panic-plan.md

```diff
@@ -0,0 +1,187 @@
+  that still allows programs with std to panic with arbitrary (`Any + Send`) payloads.
+
+- We [add a lint](https://github.com/rust-lang/rust/pull/78088) for Rust 2015/2018 that warns about problem 1, 2, and 8,
+  similar to [what Clippy already has](https://rust-lang.github.io/rust-clippy/master/index.html#panic_params).
+
+  Note that this lint isn't just to warn about incompatibilities with Rust 2021, but also to warn about usages of `panic!()` that are likely mistakes.
+
+  This lint suggests add an argument to `panic!("hello: {}")`, or to insert `"{}", ` to use it literally: `panic!("{}", "hello: {}")`.
+  ([Screenshot here.](https://user-images.githubusercontent.com/783247/96643867-79eb1080-1328-11eb-8d4e-a5586837c70a.png))
+  The second suggestion can be a pessimization for code size, but I believe that [can be solved separately](https://github.com/rust-lang/rust/issues/78356).
+
+- After `panic_any` is stable, we add a lint for Rust 2015/2018 (or extend the one above) to warn about problem 3, 4, 5 and 9.
+  It warns about `panic!(x)` for anything other than a string literal, and suggests to use
+ `panic_any(x)` instead of `std::panic!(x)`, and
+ `panic!("{}", x)` instead of `core::panic!(x)`.
+
+  It will also detect problem 6 (e.g. `assert!(true, false)`) because that expands to such a panic invocation,
+  but will suggest `assert_eq!()` for this case instead.
+
+- We [modify the panic glue between core and std](https://github.com/rust-lang/rust/pull/78119)
+  to use `Arguments::as_str()` to make sure both `std::panic!("literal")` and `core::panic!("literal")`
+  result in a `&'static str` payload. This removes one of the differences between the two macros in Rust 2015/2018.
+
+  This is already merged.
+
+- Now that `std::panic!("literal")` and `core::panic!("literal")` behave identically,
+  [we modify `todo!()`, `unimplemented!()`, `assert_eq!()`, etc.](https://github.com/rust-lang/rust/pull/78343)
+  to use `$crate::panic!()` instead of `panic!()`. This solves problem 7 for all macros except `assert!()`.
+
+- We modify `assert!()` to use `$crate::panic!()` instead of `panic!()` for the single argument case in Rust 2015/2018,
+  and for all cases in Rust 2021.
+
+  This solves problem 7 for the common case of `assert!(expr)` in Rust 2015/2018, and for all cases of `assert!` in Rust 2021.
+
+Together, these actions address all problems, without breaking any existing code.
+
+# Drawbacks
+
+- This results in subtle differences between Rust editions.
+
+- This requires `assert!` and `panic!` to behave differently depending on the Rust edition of the crate it is used in.
+  `panic!` is just a `macro_rules` macro right now, which does not natively support that.
+
+# Alternatives
+
+- Instead of the last step, we could also simply break `assert!(expr, non_string_literal)` in all editions.
+  This usage is probably way less common than `panic!(non_string_literal)`.
```

Comment:
Trying to do a crater run for this failed before it even started, because that change didn't even get past the `@bors try` run. It caused a compilation error while compiling this part of `clap`:

https://github.com/clap-rs/clap/blob/0c7da9f5b32bcd6968a70258a4868d439fbc1fc3/src/app/parser.rs#L181-L184

Apparently this behaviour of `assert!()` is relied upon.